File: kalitech-engine/src/main/resources/kalitech/engine/modules/Light/Light.js
// FILE: resources/kalitech/engine/modules/Light/Light.js
// Author: Kalitech
"use strict";

const {requireEngineApi, normalizeCfgObject} = require("../helpers/ModuleCommon.js");

/**
 * Light API wrapper.
 * Provides safe helpers for creating and managing light handles.
 */
function create(engine /*, K */) {
    const api = requireEngineApi(engine, "light", "LIGHT");

    return Object.freeze({
        /** Create a new light and return its handle. */
        create(cfg) {
            return api.create(normalizeCfgObject(cfg));
        },

        /** Update an existing light. */
        set(handle, cfg) {
            api.set(handle, normalizeCfgObject(cfg));
            return handle;
        },

        /** Enable or disable a light. */
        enable(handle, enabled) {
            api.enable(handle, !!enabled);
            return handle;
        },

        /** Check if a light handle exists. */
        exists(handle) {
            return !!api.exists(handle);
        },

        /** Destroy a light handle. */
        destroy(handle) {
            api.destroy(handle);
            return true;
        },

        /** Get light configuration for a handle. */
        get(handle) {
            return api.get(handle);
        },

        /** List all lights. */
        list() {
            return api.list();
        },

        /** Access to the raw host API. */
        api
    });
}

create.META = {
    moduleId: "light",
    globalName: "LIGHT",
    version: "1.0.0",
    description: "Light wrapper for create/set/enable/destroy operations",
    engineMin: "0.1.0"
};

module.exports = create;

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Light/Light.lua
local M = {}
local lua_require_result_0 = require("../helpers/ModuleCommon.lua")
requireEngineApi = lua_require_result_0.requireEngineApi
normalizeCfgObject = lua_require_result_0.normalizeCfgObject
create = setmetatable(
    {},
    {__call = function(lua_, self, engine)
        local api = requireEngineApi(_G, engine, "light", "LIGHT")
        return KObject:freeze({
            create = function(self, cfg)
                return api:create(normalizeCfgObject(_G, cfg))
            end,
            set = function(self, handle, cfg)
                api:set(
                    handle,
                    normalizeCfgObject(_G, cfg)
                )
                return handle
            end,
            enable = function(self, handle, enabled)
                api:enable(handle, not not enabled)
                return handle
            end,
            exists = function(self, handle)
                return not not api:exists(handle)
            end,
            destroy = function(self, handle)
                api:destroy(handle)
                return true
            end,
            get = function(self, handle)
                return api:get(handle)
            end,
            list = function(self)
                return api:list()
            end,
            api = api
        })
    end}
)
create.META = {
    moduleId = "light",
    version = "1.0.0",
    description = "Light wrapper for create/set/enable/destroy operations",
    engineMin = "0.1.0"
}
M = create

return M
